### PR TITLE
Challenge/count vowels

### LINKED ISSUE #3: Count Vowels
| **Field**        | **Details**                                                                                                         |
|------------------|---------------------------------------------------------------------------------------------------------------------|
| **Name**         | New Challenge                                                                                                       |
| **Description**        | Write a program that counts the number of vowels in a given string.                                 |
| **Title**        | Count Vowels                                                                                                        |

### PR DESCRIPTION
**Pull request by danielluyi**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/count_vowels.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+def count_vowels(input_string):
```
Comment (NoorelsalamAlmakki) [inline solutions/count_vowels.py:1]: Add module header.
```suggestion
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
A module for calculating the <description of the behavior of the function>

Module contents:
    - <function name>: <description of the behavior of the function>.

Created on <date created>
@author: <Your Name>
"""
```

File: solutions/count_vowels.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    Returns:
+        int: The total number of vowels (a, e, i, o, u) in the string,
+             including both uppercase and lowercase vowels.
+
+    Example:
+        >>> count_vowels("Hello, World!")
+        3
+        >>> count_vowels("Python")
+        1
```
Comment (NoorelsalamAlmakki) [inline solutions/count_vowels.py:16]: Add more examples; 3 to 5 is enough.

File: solutions/count_vowels.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    for char in input_string:
+        if char in vowels:
+            count += 1
+    return count
+
+
+# Example usage
+if __name__ == "__main__":
+    input_string = "Hello, how many vowels are here?"
+    print("Number of vowels:", count_vowels(input_string))
```
Comment (NoorelsalamAlmakki) [inline solutions/count_vowels.py]: The function file should not contain any print() statement.

File: solutions/count_vowels.py (hunk 2 above)
Comment (NoorelsalamAlmakki) [inline solutions/count_vowels.py]: Add more examples; 3 to 5 is enough.

File: solutions/count_vowels.py (hunk 1 above)
Comment (NoorelsalamAlmakki) [inline solutions/count_vowels.py]: Add a module header. The following is an example, replace the text between <> with the appropriate input.
```suggestion
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
A module for calculating the <description of the behavior of the function>
Module contents:
    - <function name>: <description of the behavior of the function>.
Created on <date created>
@author: <Your Name>
"""

def count_vowels(input_string):
```

File: solutions/tests/test_count_vowels.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import unittest
```
Comment (NoorelsalamAlmakki) [inline solutions/tests/test_count_vowels.py]: The same applies here. Add a module docstring. It's the same as the previous one. However, you need to change the description so it reflects that this is a module to test the function <function name>

File: solutions/count_vowels.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    Returns:
+        int: The total number of vowels (a, e, i, o, u) in the string,
+             including both uppercase and lowercase vowels.
+
+    Example:
+        >>> count_vowels("Hello, World!")
+        3
+        >>> count_vowels("Python")
+        1
+    """
```
Comment (NoorelsalamAlmakki) [inline solutions/count_vowels.py:38]: Add assertion to make sure the input value is of a string type.
```suggestion
    """
    assert isinstance(input_string, str), "The input value should be a string"
```

**Review by NoorelsalamAlmakki — CHANGES_REQUESTED** [review]
There are changes that need to be addressed before merging.

File: solutions/count_vowels.py (hunk 3 above)
Comment (danielluyi) [reply solutions/count_vowels.py]: Alright. I'm working on all the corrections at the moment

**Review by NoorelsalamAlmakki — DISMISSED** [review]
The branch is ready to be merged.

**Review by NoorelsalamAlmakki — DISMISSED** [review]
